tardis/util/base.py

```diff
@@ -596,3 +597,124 @@ def is_notebook():
+
+
+def update_packet_pbar(i, current_iteration, no_of_packets, total_iterations):
+    """
+    Update progress bars as each packet is propagated.
+
+    Parameters
+    ----------
+    i : int
+        Amount by which the progress bar needs to be updated.
+    current_iteration : int
+        Current iteration number.
+    no_of_packets : int
+        Total number of packets in one iteration.
+    total_iterations : int
+        Total number of iterations.
+    """
+    if packet_pbar.desc == "":
+        packet_pbar.desc = "0"
+    bar_iteration = int(packet_pbar.desc) - 1
+
+    # fix bar layout when run_tardis is called for the first time
+    if iterations_pbar.total == None:
+        fix_bar_layout(iterations_pbar, total_iterations=total_iterations)
+    if packet_pbar.total == None:
+        fix_bar_layout(packet_pbar, no_of_packets=no_of_packets)
+
+    # display and reset progress bar when run_tardis is called again
+    if iterations_pbar.n == total_iterations:
+        if type(iterations_pbar).__name__ == "tqdm_notebook":
+            iterations_pbar.container.close()
+        fix_bar_layout(iterations_pbar, total_iterations=total_iterations)
+
+    if bar_iteration > current_iteration:
+        packet_pbar.desc = current_iteration
+        if type(packet_pbar).__name__ == "tqdm_notebook":
+            # stop displaying last container
+            packet_pbar.container.close()
+        fix_bar_layout(packet_pbar, no_of_packets=no_of_packets)
+
+    # reset progress bar with each iteration
+    if bar_iteration < current_iteration:
+        packet_pbar.reset(total=no_of_packets)
+        packet_pbar.desc = str(current_iteration + 1)
+
+    packet_pbar.update(i)
+
+
+def update_iterations_pbar(i):
+    """
+    Update progress bar for each iteration.
+
+    Parameters
+    ----------
+    i : int
+        Amount by which the progress bar needs to be updated.
+    """
+    iterations_pbar.update(i)
+
+
+def fix_bar_layout(bar, no_of_packets=None, total_iterations=None):
+    """
+    Fix the layout of progress bars.
+
+    Parameters
+    ----------
+    bar : tqdm instance
+        Progress bar to change the layout of.
+    no_of_packets : int, optional
+        Number of packets to be propagated.
+    total_iterations : int, optional
+        Total number of iterations.
+    """
+    if type(bar).__name__ == "tqdm_notebook":
+        bar.container = bar.status_printer(
+            bar.fp,
+            bar.total,
+            bar.desc,
+            bar.ncols,
+        )
+        if no_of_packets is not None:
+            bar.reset(total=no_of_packets)
+        if total_iterations is not None:
+            bar.reset(total=total_iterations)
+
+        # change the amount of space the prefix string of the bar takes
+        # here, either packets or iterations
+        bar.container.children[0].layout.width = "6%"
+
+        # change the length of the bar
+        bar.container.children[1].layout.width = "60%"
```

Comment:
Have you stress tested these settings with very narrow and very wide windows for browser and terminal?

Reply:
For the notebook, these settings cause the bar to have a fixed length _relative_ to the length of the cell, meaning the actual bar would always be 60% the length of the cell and the string in front of the bar would always be 6% the length of the cell. These work great under normal conditions, but not ideally under _too_ narrow conditions, where the bar overlaps the string before the bar sometimes.

For the terminal, I defined the settings now [here](https://github.com/tardis-sn/tardis/pull/1703/files#diff-389972ec71fe92482e51ea9a059cc62a97a39f6cd7429575c3cbf63cbe3d94b2R618), which is makes sure that the string in the front of the bar is not cut short (and are equal) and the both the bars have only [80 characters](https://github.com/tqdm/tqdm#description-and-additional-stats), making sure they are of the same length. 

I hard coded the values for the terminal and they work pretty well. I could could code it in a way to always find out somehow how big the screen is and increase the length of the bars, but I didn't think that was necessary, as the user could always increase/decrease the size of the terminal text.